src/librustc_mir/const_eval.rs

```diff
@@ -93,15 +72,36 @@ fn op_to_const<'tcx>(
         _ => false,
     };
     let normalized_op = if normalize {
-        Err(*ecx.read_immediate(op).expect("normalization works on validated constants"))
+        Err(ecx.read_immediate(op).expect("normalization works on validated constants"))
     } else {
+        // FIXME(oli-obk): if we have an `Immediate::ScalarPair` here, that seems bad as
+        // we only want to allow slices.
         op.try_as_mplace()
     };
     let val = match normalized_op {
-        Ok(mplace) => return mplace_to_const(ecx, mplace),
-        Err(Immediate::Scalar(x)) =>
+        Ok(mplace) => {
+            let ptr = mplace.ptr.to_ptr()?;
+            let alloc = ecx.tcx.alloc_map.lock().unwrap_memory(ptr.alloc_id);
+            if alloc.align == mplace.align {
+                // we did not read e.g. a `(u32, u64, u8, u32)` from a packed struct.
+                // thus, we don't need to adjust the allocation's alignment
+                ConstValue::ByRef(ptr, alloc)
+            } else {
+                // we are creating a new allocation and pointer to make sure that further
+                // accesses to the returned constant don't error out because of alignment checks
+                // FIXME: this is only needed for pattern matching and simd_shuffle, can
+                // we be more permissive there and just ignore alignment? That would save
+                // us from duplicating this allocation.
```

Comment:
This seems new? Why is this needed now but was not needed before?

Reply:
We've been doing this unconditionally in `op_to_const` on all constants. You asked for the double interning to be gone in this PR, so I got rid of it where possible. We still need it for the explanation above.

Reply:
Wait, but you said that we can `assert!(alloc.align >= mplace.align)`. How can a *too strongly aligned* allocation be a problem? A bigger `alloc.align` can never make alignment checks fail, right?

Reply:
If you pattern match a packed struct, you'll end up with a `ByRef` to an inner field, but when you later access that inner field (which you now don't know the alignment of anymore), you'll try to access it with the type's alignment. The type's alignment may be bigger than `1`, which is the alignment of the memory due to the packedness

Reply:
Oh. this check might be wrong. Huh... I'll try to figure out what's goin on

Reply:
Any update on this?